torch/csrc/autograd/python_function.cpp

```diff
@@ -411,140 +503,183 @@ static void _mark_non_differentiable(THPFunction *self, t2var_type &t2var)
-        } else {
-          prev_fn = input_var->cdata;
-        }
-        self->cdata.previous_functions[i] = std::make_pair<>(prev_fn, input_var->cdata->output_nr);
-      }
+      THPVariable* variable = (THPVariable*)arg;
+      new_arg = THPVariable_get_data(variable);
+      unpacked.input_vars.push_back(variable->cdata);
+      PyObject* needs_grad = variable->cdata->requires_grad ? Py_True : Py_False;
+      Py_INCREF(needs_grad);
+      PyTuple_SET_ITEM(flags.needs_input_grad.get(), i, needs_grad);
+    }
+    PyTuple_SET_ITEM(unpacked.tensor_input.get(), i, new_arg);
+  }
 
-      std::unordered_set<PyObject *> dirty_inputs;
-      _mark_dirty(self, t2var, dirty_inputs);
-      _wrap_outputs(self, t2var, dirty_inputs, raw_output, outputs);
-      _join_version_counters(self, t2var);
-      if (self->cdata.requires_grad || self->cdata.is_stochastic) {
-        _save_variables(self, t2var);
-        _mark_non_differentiable(self, t2var);
-      } else {
-        // Remove unnecessary attributes
-        Py_XDECREF(self->to_save);
-        self->to_save = NULL;
-        Py_XDECREF(self->non_differentiable);
-        self->non_differentiable = NULL;
-      }
+  flags.flags = Function::flags(unpacked.input_vars);
+  return std::make_pair(std::move(unpacked), std::move(flags));
+}
+
+PyObject* process_outputs(THPFunction* grad_fn, const UnpackedInput& unpacked, THPObjectPtr raw_output) {
+  bool unpack_output = _ensure_tuple(raw_output);
+
+  auto num_outputs = PyTuple_GET_SIZE(raw_output.get());
+
+  THPObjectPtr outputs = PyTuple_New(num_outputs);
+  if (!outputs) throw python_error();
+  if (!grad_fn) { // if volatile
+    // If one of the inputs is volatile let's take a fast path - we want
+    // minimize the overhead of inference
+    for (int i = 0; i < num_outputs; i++) {
+      PyObject *output = PyTuple_GET_ITEM(raw_output.get(), i);
+      THPVariable *output_var = (THPVariable*)THPVariable_NewVolatile(output);
+      if (!output_var) throw python_error();
+      output_var->cdata->output_nr = i;
+      PyTuple_SET_ITEM(outputs.get(), i, (PyObject*)output_var);
+    }
+  } else {
+    grad_fn->cdata.num_inputs = num_outputs;
+
+    // Initialize t2var map
+    t2var_type t2var;
+    for (auto& c_var : unpacked.input_vars) {
+      THPVariable* py_var = (THPVariable*)c_var->pyobj;
+      t2var.emplace(py_var->data, py_var);
     }
 
-    // Unpack the output, unless .forward() returned a tuple
-    if (unpack_output) {
-      PyObject *output = PyTuple_GET_ITEM(outputs.get(), 0);
-      Py_INCREF(output);
-      return output;
+    std::unordered_set<PyObject *> dirty_inputs;
+    _mark_dirty(grad_fn, t2var, dirty_inputs);
+    _wrap_outputs(grad_fn, t2var, dirty_inputs, raw_output, outputs);
+    _join_version_counters(grad_fn, t2var);
+    // TODO: is this stochastic check needed here?
+    if (grad_fn->cdata.is_executable || grad_fn->cdata.is_stochastic) {
+      _save_variables(grad_fn, t2var);
```

Comment:
This should be performed after marking the variables as non differentiable (next line) otherwise the saved variables will always have `requires_grad=True` (trying to accumulate their gradient will fail as they don't have a grad accumulator and will crash).